### ISSUE TITLE
Lazy decomposition wrong argument order

### ISSUE THREAD

**Issue by thomasaarholt** [hyperspy/hyperspy#1589]
The order the first three arguments for decomposition is confusing. I understand why it is necessary that the argument with no default value is first, but it would be nice if there were any good suggestions to fix this.

I also just discovered that lazy decomposition proceeds when output dimension= True, which gives no meaning. 

```python
# Normal signals:
s.decomposition(normalize_poissonian_noise=False, algorithm='svd', output_dimension=None, centre=None, auto_transpose=True, navigation_mask=None, signal_mask=None, var_array=None, var_func=None, polyfit=None, reproject=None, return_info=False, **kwargs)
#Lazy signals:
s.decomposition(output_dimension, normalize_poissonian_noise=False, algorithm='PCA', signal_mask=None, navigation_mask=None, get=<function get at 0x0000000009319E18>, num_chunks=None, reproject=True, bounds=True, **kwargs)

**Comment by to266** [comment]
`int(True)` is just `1`, so I assume lazy decomposition with one component was performed. This is meaningful, albeit likely just the mean of the data.

As you already said, the argument without default __has__ to be first. A few suggestions:

 - Either the users should pay more attention and not rely on the order of the argument (i.e. pass kwargs explicitly unless you are able to figure out the problem if the order of the kwargs changes)
 - We change the long-standing API of normal signals and replicate the lazy signal argument order (which is necessary). This would likely break __many__ notebooks, where people did not follow the first suggestion. They would not be able to figure out the problem, though.
 - We check that `output_dimension` passed to the lazy signal is explicitly integer (so that it fails if `True` is passed). While this would likely work, I think it's a horrible fix for such a problem that should be solved by encouraging users to learn more python.

Overall I think it's not a hyperspy issue

**Comment by thomasaarholt** [comment]
I fully agree with the comments above. Perhaps we can require that the `type(output_dimension) == int` so that silly people like me get warned when they accidentally enter `output_dimension = True` by `s.decomposition(True)`?

**Comment by francisco-dlp** [comment]
+1 for a more informative error and these will likely become a recurrent issue.

**Comment by francisco-dlp** [comment]
Fixed in #1878

**Comment by ericpre** [comment]
#1878 is a PR to next minor, should the milestone of this issue changed to 1.4?

**Comment by francisco-dlp** [comment]
:+1: , also there is no way to get this sorted by tomorrow.

**Comment by francisco-dlp** [comment]
Fixed in #1918
